File: sdks/swift/Sources/AstroSDK/Presentments/PresentmentsClient.swift
import Foundation

public final class PresentmentsClient {
    private let http: AstroHTTPClient
    init(http: AstroHTTPClient) { self.http = http }

    public func capabilities() async throws -> PresentedPaymentCapabilitiesResponse {
        let req = http.request(http.url("/capabilities"))
        return try await http.perform(req)
    }

    public func create(_ request: CreatePresentmentRequest, idempotencyKey: String? = nil) async throws -> Presentment {
        let headers = idempotencyKey.map { ["Idempotency-Key": $0] } ?? [:]
        let req = http.request(http.url("/presentments"), method: "POST", body: request, extraHeaders: headers)
        return try await http.perform(req)
    }

    public func list() async throws -> [Presentment] {
        let req = http.request(http.url("/presentments"))
        let response: PresentmentsListResponse = try await http.perform(req)
        return response.data
    }

    public func get(_ presentmentId: String) async throws -> Presentment {
        let req = http.request(http.url("/presentments/\(presentmentId)"))
        return try await http.perform(req)
    }

    public func claim(_ presentmentId: String, request: ClaimPresentmentRequest) async throws -> Presentment {
        let req = http.request(http.url("/presentments/\(presentmentId)/claim"), method: "POST", body: request)
        return try await http.perform(req)
    }

    public func cancel(_ presentmentId: String) async throws -> Presentment {
        let req = http.request(http.url("/presentments/\(presentmentId)/cancel"), method: "POST")
        return try await http.perform(req)
    }

    public func status(_ presentmentId: String) async throws -> PresentmentStatusResponse {
        let req = http.request(http.url("/presentments/\(presentmentId)/status"))
        return try await http.perform(req)
    }
}

File: sdks/kotlin/src/main/kotlin/ly/neptune/astro/presentments/PresentmentsClient.kt
package ly.neptune.astro.presentments

import ly.neptune.astro.HttpEngine
import ly.neptune.astro.models.*

class PresentmentsClient internal constructor(private val engine: HttpEngine) {

    suspend fun capabilities(): PresentedPaymentCapabilitiesResponse =
        engine.get("/capabilities")

    suspend fun create(request: CreatePresentmentRequest, idempotencyKey: String? = null): Presentment =
        engine.post(
            "/presentments",
            request,
            idempotencyKey?.let { mapOf("Idempotency-Key" to it) } ?: emptyMap()
        )

    suspend fun get(presentmentId: String): Presentment =
        engine.get("/presentments/$presentmentId")

    suspend fun list(): List<Presentment> =
        engine.get<PresentmentsListResponse>("/presentments").data

    suspend fun claim(presentmentId: String, request: ClaimPresentmentRequest): Presentment =
        engine.post("/presentments/$presentmentId/claim", request)

    suspend fun cancel(presentmentId: String): Presentment =
        engine.post("/presentments/$presentmentId/cancel")

    suspend fun status(presentmentId: String): PresentmentStatusResponse =
        engine.get("/presentments/$presentmentId/status")
}
